core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -10180,6 +10214,34 @@ private AuthenticationResult generateUserStoreChainWithID(String preferredUserNa
             String preferredUserNameValue, Object credential, String profileName, boolean domainProvided,
             List<String> userStorePreferenceOrder) throws UserStoreException {
 
+        // If domain name is provided, directly authenticate using the corresponding user store.
+        if (domainProvided) {
+            String providedDomainName = this.getUserStore(preferredUserNameValue).getDomainName();
```

Comment:
Why don't we use `UserCoreUtil.extractDomainFromName(userName)` here.

Reply:
'preferredUserNameValue' can be a different claim other than the username.